### ISSUE TITLE
Improve error messages from kind checker

### ISSUE THREAD

**Issue by puffnfresh** [purescript/purescript#1028]
I keep putting `()` instead of `Unit` and getting an error like this:

```
  Error in data binding group:
  Error checking kind of type
    Key ()
  Cannot unify kind
    *
  with kind
    # u16
```

It'd be great if we could special case whether `()` was used and add:

> Maybe you meant to use `Unit`.

**Comment by garyb** [comment]
:+1:

**Comment by paf31** [comment]
:+1:

**Comment by paf31** [comment]
Renamed to "Improve error messages from kind checker", since I think we can generally improve the error reporting by doing something closer to what we do in the type checker. If the error message said something like `while checking that type () has kind *`, I think it would be more obvious what went wrong here, for example.

**Comment by puffnfresh** [comment]
:+1:

**Comment by paf31** [comment]
@natefaubion Does the new kind checker fix this?

**Comment by natefaubion** [comment]
Fix it in which way? The kind checker keeps a trace like the normal type checker, but I thought it did that before.

```
foreign import data Foo :: Type -> Type

test :: Int -> Foo ()
test = ?a
```
```
Error found:
in module Test
at Test.purs:14:21 - 14:22 (line 14, column 21 - line 14, column 22)

  Could not match kind

    Row t0

  with kind

    Type


while checking that type ()
  has kind Type
while inferring the kind of Foo ()
while inferring the kind of Int -> Foo ()
in value declaration test
```

**Comment by paf31** [comment]
I would say `while checking that type () has kind Type` is good enough.

**Comment by hdgarrood** [comment]
Let's close this once #3779 is merged then?

**Comment by i-am-the-slime** [comment]
Should this be closed then #3779 has been merged for a while.
